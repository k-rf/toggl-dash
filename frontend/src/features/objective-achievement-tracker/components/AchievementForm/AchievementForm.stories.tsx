import type { Meta, StoryObj } from "@storybook/react";
// import { userEvent, waitFor, within } from "@storybook/testing-library";

import { AchievementForm } from "./AchievementForm";

export default { component: AchievementForm } as Meta<typeof AchievementForm>;

type Story = StoryObj<typeof AchievementForm>;

export const Default: Story = {
  args: {},
};
