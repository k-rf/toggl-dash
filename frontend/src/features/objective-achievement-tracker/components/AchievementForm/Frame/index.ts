export * from "./DayFieldFrame";
export * from "./HourFieldFrame";
export * from "./RowFrame";
