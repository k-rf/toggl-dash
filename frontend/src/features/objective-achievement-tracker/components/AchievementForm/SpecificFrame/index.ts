export * from "./HolidayFrame";
export * from "./OffDayFrame";
export * from "./TotalFrame";
export * from "./WeekdayFrame";
