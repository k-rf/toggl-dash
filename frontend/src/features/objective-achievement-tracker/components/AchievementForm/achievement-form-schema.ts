import { z } from "zod";

export const achievementFormSchema = z
  .object({
    workday: z.preprocess(Number, z.number().int().min(0).max(31)),
    holiday: z.preprocess(Number, z.number().int().min(0).max(31)),
    offDay: z.preprocess(Number, z.number().int().min(0).max(31)),
    workdayHour: z.preprocess(Number, z.number().int().min(0).max(24)),
    holidayHour: z.preprocess(Number, z.number().int().min(0).max(24)),
    correlation: z.undefined(),
  })
  .superRefine(({ workday, holiday, offDay }, ctx) => {
    const addIssue = (path: string, message: string) => {
      ctx.addIssue({ code: "custom", path: [path], message });
    };

    if (31 < workday + holiday + offDay) {
      for (const path of ["workday", "holiday", "offDay", "correlation"]) {
        addIssue(path, "Over 31 days");
      }
    }

    if (workday + holiday + offDay < 28) {
      for (const path of ["workday", "holiday", "offDay", "correlation"]) {
        addIssue(path, "Under 28 days");
      }
    }
  });

export type AchievementFormSchema = z.infer<typeof achievementFormSchema>;
