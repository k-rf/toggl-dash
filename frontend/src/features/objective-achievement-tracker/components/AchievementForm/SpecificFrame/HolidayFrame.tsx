import { Control } from "react-hook-form";

import { RhfTextField } from "~/components/Compositions";
import { ElTypography } from "~/components/Elements/ElTypography";
import { AvailableTimeFormSchema } from "~/features/objective-achievement-tracker/hooks";

import { RowFrame, DayFieldFrame, HourFieldFrame } from "../Frame";

type Props = {
  control: Control<AvailableTimeFormSchema>;
  isDayError?: boolean;
  isHourError?: boolean;
};

export const HolidayFrame = (props: Props) => {
  return (
    <RowFrame>
      <ElTypography>休日</ElTypography>
      <DayFieldFrame>
        {(commonProps) => (
          <RhfTextField
            control={props.control}
            defaultValue={0}
            name="holiday"
            error={props.isDayError}
            {...commonProps}
          />
        )}
      </DayFieldFrame>
      <HourFieldFrame>
        {(commonProps) => (
          <RhfTextField
            control={props.control}
            defaultValue={0}
            name="holidayHour"
            error={props.isHourError}
            {...commonProps}
          />
        )}
      </HourFieldFrame>
    </RowFrame>
  );
};
