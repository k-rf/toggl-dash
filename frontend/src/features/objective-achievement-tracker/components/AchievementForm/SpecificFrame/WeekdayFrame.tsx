import { Control } from "react-hook-form";

import { RhfTextField } from "~/components/Compositions";
import { ElTypography } from "~/components/Elements/ElTypography";
import { AvailableTimeFormSchema } from "~/features/objective-achievement-tracker/hooks";

import { RowFrame, DayFieldFrame, HourFieldFrame } from "../Frame";

type Props = {
  control: Control<AvailableTimeFormSchema>;
  isDayError?: boolean;
  isHourError?: boolean;
};

export const WeekdayFrame = (props: Props) => {
  return (
    <RowFrame>
      <ElTypography>平日</ElTypography>
      <DayFieldFrame>
        {(commonProps) => (
          <RhfTextField
            control={props.control}
            defaultValue={0}
            name="weekday"
            error={props.isDayError}
            {...commonProps}
          />
        )}
      </DayFieldFrame>
      <HourFieldFrame>
        {(commonProps) => (
          <RhfTextField
            control={props.control}
            defaultValue={0}
            name="weekdayHour"
            error={props.isHourError}
            {...commonProps}
          />
        )}
      </HourFieldFrame>
    </RowFrame>
  );
};
