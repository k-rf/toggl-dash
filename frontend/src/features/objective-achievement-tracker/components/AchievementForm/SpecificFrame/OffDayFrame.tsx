import { Box, Typography } from "@mui/material";
import { Control } from "react-hook-form";

import { RhfTextField } from "~/components/Compositions";
import { ElTypography } from "~/components/Elements/ElTypography";

import { AchievementFormSchema } from "../achievement-form-schema";
import { RowFrame, DayFieldFrame } from "../Frame";

type Props = {
  control: Control<AchievementFormSchema>;
  isDayError?: boolean;
};

export const OffDayFrame = (props: Props) => {
  return (
    <RowFrame>
      <ElTypography>オフ</ElTypography>
      <DayFieldFrame>
        {(commonProps) => (
          <RhfTextField
            control={props.control}
            defaultValue={0}
            name="offDay"
            error={!!props.isDayError}
            {...commonProps}
          />
        )}
      </DayFieldFrame>
      <Box display="flex" gap={(theme) => theme.spacing(0, 1)}>
        <Typography sx={{ width: 64, textAlign: "end" }}>0</Typography>
        <Typography>時間</Typography>
      </Box>
    </RowFrame>
  );
};
