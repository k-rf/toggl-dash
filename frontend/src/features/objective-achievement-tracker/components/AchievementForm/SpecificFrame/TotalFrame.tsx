import { Box, Typography } from "@mui/material";
import { UseFormWatch } from "react-hook-form";

import { ElTypography } from "~/components/Elements/ElTypography";

import { AchievementFormSchema } from "../achievement-form-schema";
import { RowFrame } from "../Frame";

type Props = {
  watch: UseFormWatch<AchievementFormSchema>;
};

export const TotalFrame = (props: Props) => {
  const [workday, holiday, offDay, workdayHour, holidayHour] = (
    ["workday", "holiday", "offDay", "workdayHour", "holidayHour"] as const
  ).map((e) => props.watch(e, 0)) as [number, number, number, number, number];

  return (
    <RowFrame>
      <ElTypography>合計</ElTypography>
      <Box display="flex" gap={(theme) => theme.spacing(0, 1)}>
        <Typography sx={{ width: 48, textAlign: "end" }}>
          {sum([workday, holiday, offDay])}
        </Typography>
        <Typography>日</Typography>
      </Box>
      <Box display="flex" gap={(theme) => theme.spacing(0, 1)}>
        <Typography sx={{ width: 64, textAlign: "end" }}>
          {sum([workday * workdayHour, holiday * holidayHour])}
        </Typography>
        <Typography>時間</Typography>
      </Box>
    </RowFrame>
  );
};

const sum = (values: (number | string)[]) => {
  return values.map(Number).reduce((p, c) => p + c, 0);
};
