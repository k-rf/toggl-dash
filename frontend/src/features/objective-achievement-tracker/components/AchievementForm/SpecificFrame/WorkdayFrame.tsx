import { Control } from "react-hook-form";

import { RhfTextField } from "~/components/Compositions";
import { ElTypography } from "~/components/Elements/ElTypography";

import { AchievementFormSchema } from "../achievement-form-schema";
import { RowFrame, DayFieldFrame, HourFieldFrame } from "../Frame";

type Props = {
  control: Control<AchievementFormSchema>;
  isDayError?: boolean;
  isHourError?: boolean;
};

export const WorkdayFrame = (props: Props) => {
  return (
    <RowFrame>
      <ElTypography>平日</ElTypography>
      <DayFieldFrame>
        {(commonProps) => (
          <RhfTextField
            control={props.control}
            defaultValue={0}
            name="workday"
            error={props.isDayError}
            {...commonProps}
          />
        )}
      </DayFieldFrame>
      <HourFieldFrame>
        {(commonProps) => (
          <RhfTextField
            control={props.control}
            defaultValue={0}
            name="workdayHour"
            error={props.isHourError}
            {...commonProps}
          />
        )}
      </HourFieldFrame>
    </RowFrame>
  );
};
