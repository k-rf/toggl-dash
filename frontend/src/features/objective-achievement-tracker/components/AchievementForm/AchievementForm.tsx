import { zodResolver } from "@hookform/resolvers/zod";
import { Stack } from "@mui/material";
import { useEffect } from "react";
import { useForm } from "react-hook-form";

import { ElTypography } from "~/components/Elements/ElTypography";

import { AchievementFormSchema, achievementFormSchema } from "./achievement-form-schema";
import { HolidayFrame, OffDayFrame, TotalFrame, WorkdayFrame } from "./SpecificFrame";

export const AchievementForm = () => {
  const {
    control,
    watch,
    trigger,
    formState: {
      isDirty,
      errors: {
        workday: workdayError,
        workdayHour: workdayHourError,
        holiday: holidayError,
        holidayHour: holidayHourError,
        offDay: offDayError,
        correlation: correlationError,
      },
    },
  } = useForm<AchievementFormSchema>({
    resolver: zodResolver(achievementFormSchema),
    mode: "all",
  });

  useEffect(() => {
    if (isDirty) trigger(["workday", "holiday", "offDay", "correlation"]);
  }, [
    trigger,
    isDirty,
    workdayError?.message,
    holidayError?.message,
    offDayError?.message,
    correlationError?.message,
  ]);

  return (
    <Stack spacing={2}>
      <WorkdayFrame
        control={control}
        isDayError={!!workdayError}
        isHourError={!!workdayHourError}
      />
      <HolidayFrame
        control={control}
        isDayError={!!holidayError}
        isHourError={!!holidayHourError}
      />
      <OffDayFrame control={control} isDayError={!!offDayError} />
      <TotalFrame watch={watch} />

      <ElTypography color={(theme) => theme.palette.error.main}>
        {correlationError?.message}
      </ElTypography>
    </Stack>
  );
};
