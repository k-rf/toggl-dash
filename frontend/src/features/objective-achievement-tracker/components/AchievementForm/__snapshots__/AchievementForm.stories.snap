// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/Objective-achievement-tracker/Components/AchievementForm/AchievementForm > Default 1`] = `
<body
  style=""
>
  <div>
    <div
      class="css-1p5q5e5-MuiStack-root"
    >
      <div
        class="css-1yo1jo6-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          平日
        </p>
        <div
          class="MuiBox-root css-njvxaa"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-ftoozn-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id="workday"
                max="31"
                min="0"
                name="workday"
                type="number"
                value="0"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            日
          </p>
        </div>
        <div
          class="MuiBox-root css-njvxaa"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-l3h2ne-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id="workdayHour"
                max="24"
                min="0"
                name="workdayHour"
                type="number"
                value="0"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            時間
          </p>
        </div>
      </div>
      <div
        class="css-1yo1jo6-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          休日
        </p>
        <div
          class="MuiBox-root css-njvxaa"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-ftoozn-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id="holiday"
                max="31"
                min="0"
                name="holiday"
                type="number"
                value="0"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            日
          </p>
        </div>
        <div
          class="MuiBox-root css-njvxaa"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-l3h2ne-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id="holidayHour"
                max="24"
                min="0"
                name="holidayHour"
                type="number"
                value="0"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            時間
          </p>
        </div>
      </div>
      <div
        class="css-1yo1jo6-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          オフ
        </p>
        <div
          class="MuiBox-root css-njvxaa"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-ftoozn-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id="offDay"
                max="31"
                min="0"
                name="offDay"
                type="number"
                value="0"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            日
          </p>
        </div>
        <div
          class="MuiBox-root css-t0xqtr"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-110lie9-MuiTypography-root"
          >
            0
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            時間
          </p>
        </div>
      </div>
      <div
        class="css-1yo1jo6-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          合計
        </p>
        <div
          class="MuiBox-root css-t0xqtr"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-fy8r4z-MuiTypography-root"
          >
            0
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            日
          </p>
        </div>
        <div
          class="MuiBox-root css-t0xqtr"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-110lie9-MuiTypography-root"
          >
            0
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            時間
          </p>
        </div>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1wc1u5z-MuiTypography-root"
      />
    </div>
  </div>
</body>
`;
